File: leet_MONOSTACK_0962_max_width.py
class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        Q = collections.deque()
        # idea/ - we need a monotonic decreasing stack
        for i, n in enumerate(nums):
            if not Q or n < nums[Q[0]]:
                Q.appendleft(i)
        res = 0
        for i in range(len(nums) - 1, -1, -1):
            while Q and nums[Q[0]] <= nums[i]:
                res = max(res, i - Q[0])
                Q.popleft()
        return res

File: leet_MONOSTACK_0962_max_width.go
func maxWidthRamp(nums []int) int {
    Q := []int{}
    N := len(nums)
    i := 0
    for i < N {
        if len(Q) == 0 || nums[Q[ len(Q) - 1 ]] > nums[i] {
            Q = append(Q, i)
        }
        i++
    }
    i = N - 1
    res := 0
    for i > -1 {
        for len(Q) > 0 && nums[Q[ len(Q) - 1 ]] <= nums[i] {
            if res < i - Q[len(Q) - 1] {
                res = i - Q[len(Q) - 1]
            }
            Q = Q[:len(Q) - 1]
        }
        i--
    }
    return res
}
